File: motion_from_pixel_distance/script/motion.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include "ros/ros.h"
#include <math.h>
#include <geometry_msgs/PoseStamped.h>
#include <std_msgs/String.h>

geometry_msgs::Twist vel;
geometry_msgs::Twist vel_linear;
geometry_msgs::Twist vel_angular;
void split(const string& s,vector<int>& sv,const char flag = ' ') {
    sv.clear();
    istringstream iss(s);
    string temp;
    while (getline(iss, temp, flag)) {
        sv.push_back(stoi(temp));
    }
    return;
}

void chatterCallback(const std_msgs::String::ConstPtr& msg)
{
    //pixel_distance_x,pixel_distance_y
    vector<int> sv;
    int pixel_distance_x,pixel_distance_y,angularz;

    split(msg, sv, ',');
    pixel_distance_x = sv[0];
    pixel_distance_y = sv[1];

    // linearx
    if (pixel_distance_y > 5)
        vel.linear.x = 0.3;
    else if (pixel_distance_y < -5)
        vel.linear.x = -0.3;
    else 
        vel.linear.x = 0.0;

    //angularz
    if (pixel_distance_x >-5 && pixel_distance_x < 5)
        angularz = 0;
    else 
        angularz = atan2(pixel_distance_y/pixel_distance_x);
    vel.angular.z = angularz;

    cmd_pub.publish(vel);
}
 
int main(int argc, char **argv)
{
    ros::init(argc,argv,"motion");
    ros::NodeHandle n;
    ros::Subscriber sub = n.subscribe("/pixel_distance",1,chatterCallback);
    ros::Publisher cmd_pub = nhandle.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
    ros::Publisher pub_angular = nhandle.advertise<geometry_msgs::Twist>("/pixel_distance/steering", 1);
    ros::Publisher pub_linear = nhandle.advertise<geometry_msgs::Twist>("/pixel_distance/steering", 1);

    ros::Rate loop_rate(10);
    while(ros::ok())
    {
        cmd_pub.publish(vel);
        pub_angular.publish(vel);
        pub_linear.publish(vel);
        ros::spinOnce();
        loop_rate.sleep();
    }
    return 0;
}

File: motion_from_pixel_distance/script/motion.py
#!/usr/bin/env python
from __future__ import print_function

import roslib
import sys
import rospy
import re

from std_msgs.msg import String
from geometry_msgs.msg import Twist


class motion:

  def __init__(self):

    self.cmd_vel_pub = rospy.Publisher('/cmd_vel', Twist, queue_size=1)
    self.twist = Twist()

    self.string_sub_steering = rospy.Subscriber("/pixel_distance/steering", String, self.callback_steering)
    self.string_sub_running = rospy.Subscriber("/pixel_distance/running", String, self.callback_running)

  def callback_steering(self, msg):
    # receive type String, "12,89"
    err_string = msg.data
    # split int from string, "12,89"
    err_ints = re.split(',', err_string);

    err_int_x = err_ints[0]

    self.twist.linear.x = 0
    self.twist.angular.z = -float(err_int_x) / 100
    self.cmd_vel_pub.publish(self.twist)
    print(self.twist)

    #
    # # BEGIN CONTROL
    # err = cx - 1*w/2
    # self.twist.linear.x = 1
    # dt = rospy.get_time() - ptime
    # self.twist.angular.z = (-float(err) / 100)*1 + ((err - perr)/(rospy.get_time() - 		ptime))*1/15/100 #+ (serr*dt)*1/20/100 #1 is best, starting 3 unstable
    # serr = err + serr
    # perr = err
    # ptime = rospy.get_time()
    #
    # self.cmd_vel_pub.publish(self.twist)

  def callback_running(self, msg):
       # receive type String, "12,89"
       err_string = msg.data
       # split int from string, "12,89"
       err_ints = re.split(',', err_string);

       err_int_x = err_ints[0]

       self.twist.linear.x = 0.2
       self.twist.angular.z = -float(err_int_x) / 100
       self.cmd_vel_pub.publish(self.twist)
       print(self.twist)


def main(args):
  rospy.init_node('motion_from_pixel_distance', anonymous=True)
  m = motion()
  try:
    rospy.spin()
  except KeyboardInterrupt:
    print("Shutting down")


if __name__ == '__main__':
    main(sys.argv)
